### ISSUE TITLE
spread.banana should raise BananaError when receiving pb messages without pb being the selected dialect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7662]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7662|
|Type|defect|
|Created|2014-10-04 07:11:16Z|
|Branch|https://github.com/twisted/twisted/tree/banana-receive-wrong-dialect-7662|

with this patch it does

Attachments:

* [7662.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4ec/4ecd15b3015c937544de753ffa2108336b2936e4/bdeb66f06eb37a52ab490887ade67eb157187ad4.patch) (2171 bytes) - added by wolfgang61 on 2014-10-04 07:13:59Z - 
* [7662b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4ec/4ecd15b3015c937544de753ffa2108336b2936e4/566883124f86925740142dadf5aa197ba980ba14.patch) (4796 bytes) - added by wolfgang61 on 2014-10-04 12:34:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7662 7662
type__defect defect
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__ 
branch__branches_banana_receive_wrong_dialect_7662 branches/banana-receive-wrong-dialect-7662
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__spread spread
time__1412406676039407 1412406676039407
changetime__1412508836697609 1412508836697609
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to wolfgang61|
|-|-|

Thanks.

  1. The new test makes three different assertions.  It should be split into three separate test methods.
  1. A good style for testing protocols that involve obscure byte sequences is to use the protocol implementation to generate the byte sequence.  Using `Banana` to generate the VOCAB sequence would be ideal.  At a minimum, `b"\x10" + chr(VOCAB)` (with a comment about what \x10 is).
  1. It would be nice if the implementation of PB-specific vocabulary didn't involve adding a check for `... and dialect == "pb"` to every token type.
  1. This is an incompatible change to the protocol implementation.  Fortunately, there is a Banana specification which documents that \x87 is part of the "pb" "Profile".  Incompatible changes are allowed by the compatibility policy when the implementation is in non-compliance with the specification.  I suggest updating the specification to be more explicit about extensions to Banana made by new profiles must only be used if that profile is in use.  It's probably also worth adding a note to the implementation that its "dialect" concept maps to the specifications "profile" concept.
  1. By adding the check to see if the value byte is in `incomingVocabulary` the exception case for invalid vocabulary values is changed.  Before it was a `KeyError`.  Now it is `NotImplementedError("Invalid type byte ...")`.  A better behavior is probably to introduce a new exception type specifically for this case and raise it, instead.  It would be fine to do that as work towards a separate ticket - but if so, I think this patch should change so that `KeyError` is still raised.
  1. Feel free to create a new test case class for VOCAB tests.  Classes are a good way to organize related tests.
  1. The news fragment should use fully qualified names.  `twisted.spread.banana` not `spread.banana`.  And the module doesn't really raise the exception - be at least as specific as the class, `twisted.spread.banana.Banana`.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|wolfgang61 commented|
|-|-|

I hope this new patch addresses everything.

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|wolfgang61 set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [43214]) Branching to 'banana-receive-wrong-dialect-7662'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [43215]) Apply 7662b.patch 

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [43216]) Reflow new documentation to follow "semantic linefeed" guidelines.

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [43217]) Turn floating string literal into a class docstring and a comment.

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [43218]) Use new-style formatting.

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [43219]) misc minor doc tweaks

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [43220]) misc minor tweaks to some of the new unit tests

refs [#7662](https://github.com/twisted/twisted/issues/7662)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Thanks.  Looks good and buildbot is green now.  Merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [43224]) Merge banana-receive-wrong-dialect-7662

Author: wolfgang61, exarkun
Reviewer: exarkun
Fixes: [#7662](https://github.com/twisted/twisted/issues/7662)

Check the active Banana dialect before processing dialect-specific features in
the Banana protocol implementation.  Previously features of the PB dialect
could be used even when the None dialect was active.
